### ISSUE TITLE
one_hot tensors are channels_last but marked as contiguous

### ISSUE THREAD

**Issue by a-gn** [pytorch/pytorch#43195]
## 🐛 Bug

`torch.nn.functional.one_hot` outputs tensors in channels-last format, but `is_contiguous` returns `False` (and `is_contiguous(memory_format=torch.channels_last)` returns `True`).
This causes "mixed memory formats" warnings when multiplying this output with tensors that actually have the same memory format.

## To Reproduce

Steps to reproduce the behavior (in a Python interactive shell):

```python
>>> import torch
>>> a = torch.randint(10, (4, 12, 12))
>>> a.is_contiguous()
True
>>> b = torch.nn.functional.one_hot(a)
>>> b.shape
torch.Size([4, 12, 12, 10])
>>> b.is_contiguous()
True
>>> b.is_contiguous(memory_format=torch.channels_last)
False
>>> c = b.permute(0, 3, 1, 2)
>>> c.shape
torch.Size([4, 10, 12, 12])
>>> c.is_contiguous()
False
>>> c.is_contiguous(memory_format=torch.channels_last)
True
>>> d = torch.randint(10, c.shape)
>>> d.is_contiguous()
True
>>> e = c * d
<stdin>:1: UserWarning: Mixed memory format inputs detected while calling the operator. The operator will output channels_last tensor even if some of the inputs are not in channels_last format. (Triggered internally at  ..\aten\src\ATen\native\TensorIterator.cpp:924.)
>>> e.shape
torch.Size([4, 10, 12, 12])
>>> e.is_contiguous()
False
>>> e.is_contiguous(memory_format=torch.channels_last)
True
```

Note that `b` is actually in channels-last format, and `c` and `e` are in channels-first format, but `is_contiguous` says the opposite.

## Expected behavior

- `b.is_contiguous(memory_format=torch.channels_last)` should be `True`.
- `c.is_contiguous()` should be `True`.
- `e.is_contiguous()` should be `True`.
- The tensor product should not cause a warning.

## Environment

PyTorch version: 1.6.0
Is debug build: False
CUDA used to build PyTorch: 10.2

OS: Microsoft Windows 10 Education
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 1050
Nvidia driver version: 442.70
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.19.1
[pip3] torch==1.6.0
[pip3] torchvision==0.7.0
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               10.2.89              h74a9793_1
[conda] mkl                       2020.1                      216
[conda] mkl-service               2.3.0            py38hb782905_0
[conda] mkl_fft                   1.1.0            py38h45dec08_0
[conda] mkl_random                1.1.1            py38h47e9c7a_0
[conda] numpy                     1.19.1           py38h5510c5b_0
[conda] numpy-base                1.19.1           py38ha3acd2a_0
[conda] pytorch                   1.6.0           py3.8_cuda102_cudnn7_0    pytorch
[conda] torchvision               0.7.0                py38_cu102    pytorch

**Comment by SsnL** [comment]
But `b` is not in channel-last format...

**Comment by a-gn** [comment]
> But `b` is not in channel-last format...

Then I might have misunderstood what channels-last is, can you please clarify? `b` has shape `[4, 12, 12, 10]`, where `10` is the maximum value of `a`. Dimension 3 has been added for the one-hot encoding. I would expect that to be the "channel" dimension, since 0 is the batch axis and the next two are the spatial dimensions.

I transformed the tensors so that the product at the end works (the context was multiplying an image by a one-hot encoding of labels to compute a loss), and I was confused by this warning.

**Comment by gchanan** [comment]
Ya, the docs could definitely be clearer.  The following precisely gives the meaning, but it's pretty concise:
> Channels Last memory format is an alternative way of ordering NCHW tensors in memory preserving dimensions ordering. Channels Last tensors ordered in such a way that channels become the densest dimension (aka storing images pixel-per-pixel).
